File: firefly-example/src/main/java/com/firefly/example/reactive/coffee/store/vo/Page.java
package com.firefly.example.reactive.coffee.store.vo;

import com.firefly.$;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Pengtao Qiu
 */
public class Page<T> {

    private List<T> record = new ArrayList<>();
    private int number, size, total;

    private int pageCount, lastNumber, nextNumber;
    private boolean showPaging, next, last;

    public Page() {
    }

    public Page(List<T> record, int number, int size) {
        this.record = record;
        this.size = Math.max(size, 0);
        this.number = Math.max(number, 1);
        pageWithoutCount();
    }

    public Page(List<T> record, int total, int number, int size) {
        this.record = record;
        this.total = total;
        this.size = Math.max(size, 0);
        this.number = Math.max(number, 1);
        page();
    }

    public static String getPageSQLWithoutCount(int number, int size) {
        long offset = Math.max(number - 1, 0) * size;
        int pageSize = size + 1;
        return " limit " + offset + ", " + pageSize;
    }

    public static String getPageSQL(int number, int size) {
        long offset = Math.max(number - 1, 0) * size;
        return " limit " + offset + ", " + size;
    }

    public void pageWithoutCount() {
        last = number > 1;
        next = record.size() > size;
        showPaging = last || next;
        lastNumber = Math.max(number - 1, 1);
        nextNumber = next ? number + 1 : number;

        if (next) {
            record.remove(record.size() - 1);
        }
    }

    public void page() {
        pageCount = (total + size - 1) / size;
        last = number > 1;
        next = number < pageCount;
        showPaging = last || next;
        lastNumber = Math.max(number - 1, 1);
        nextNumber = Math.min(number + 1, pageCount);
    }

    public List<T> getRecord() {
        return record;
    }

    public void setRecord(List<T> record) {
        this.record = record;
    }

    public int getTotal() {
        return total;
    }

    public void setTotal(int total) {
        this.total = total;
    }

    public int getPageCount() {
        return pageCount;
    }

    public void setPageCount(int pageCount) {
        this.pageCount = pageCount;
    }

    public int getLastNumber() {
        return lastNumber;
    }

    public void setLastNumber(int lastNumber) {
        this.lastNumber = lastNumber;
    }

    public int getNextNumber() {
        return nextNumber;
    }

    public void setNextNumber(int nextNumber) {
        this.nextNumber = nextNumber;
    }

    public int getNumber() {
        return number;
    }

    public void setNumber(int number) {
        this.number = Math.max(number, 1);
    }

    public int getSize() {
        return size;
    }

    public void setSize(int size) {
        this.size = Math.max(size, 0);
    }

    public boolean isShowPaging() {
        return showPaging;
    }

    public void setShowPaging(boolean showPaging) {
        this.showPaging = showPaging;
    }

    public boolean isNext() {
        return next;
    }

    public void setNext(boolean next) {
        this.next = next;
    }

    public boolean isLast() {
        return last;
    }

    public void setLast(boolean last) {
        this.last = last;
    }

    @Override
    public String toString() {
        return $.json.toJson(this);
    }
}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/coffee/store/vo/Page.kt
package com.firefly.example.kotlin.coffee.store.vo

import com.firefly.kotlin.ext.common.Json
import java.util.*

/**
 * @author Pengtao Qiu
 */
class Page<T> {

    private var record: MutableList<T> = ArrayList()
    private var number: Int = 0
    private var size: Int = 0
    var total: Int = 0

    var pageCount: Int = 0
    var lastNumber: Int = 0
    var nextNumber: Int = 0
    var isShowPaging: Boolean = false
    var isNext: Boolean = false
    var isLast: Boolean = false

    constructor() {}

    constructor(record: MutableList<T>, number: Int, size: Int) {
        this.record = record
        this.size = Math.max(size, 0)
        this.number = Math.max(number, 1)
        pageWithoutCount()
    }

    constructor(record: MutableList<T>, total: Int, number: Int, size: Int) {
        this.record = record
        this.total = total
        this.size = Math.max(size, 0)
        this.number = Math.max(number, 1)
        page()
    }

    fun pageWithoutCount() {
        isLast = number > 1
        isNext = record.size > size
        isShowPaging = isLast || isNext
        lastNumber = Math.max(number - 1, 1)
        nextNumber = if (isNext) number + 1 else number

        if (isNext) {
            record.removeAt(record.size - 1)
        }
    }

    fun page() {
        pageCount = (total + size - 1) / size
        isLast = number > 1
        isNext = number < pageCount
        isShowPaging = isLast || isNext
        lastNumber = Math.max(number - 1, 1)
        nextNumber = Math.min(number + 1, pageCount)
    }

    fun getRecord(): List<T> {
        return record
    }

    fun setRecord(record: MutableList<T>) {
        this.record = record
    }

    fun getNumber(): Int {
        return number
    }

    fun setNumber(number: Int) {
        this.number = Math.max(number, 1)
    }

    fun getSize(): Int {
        return size
    }

    fun setSize(size: Int) {
        this.size = Math.max(size, 0)
    }

    override fun toString(): String {
        return Json.toJson(this)
    }

    companion object {
        fun getPageSQLWithoutCount(number: Int, size: Int): String {
            val offset = (Math.max(number - 1, 0) * size).toLong()
            val pageSize = size + 1
            return " limit $offset, $pageSize"
        }

        fun getPageSQL(number: Int, size: Int): String {
            val offset = (Math.max(number - 1, 0) * size).toLong()
            return " limit $offset, $size"
        }
    }
}